File: src/main/java/com/mattchowning/file_read_write/client/calls/PostFileCall.java
package com.mattchowning.file_read_write.client.calls;

import com.mattchowning.file_read_write.client.FileReadWriteClient;
import com.mattchowning.file_read_write.client.handler.HandlerCallback;
import com.mattchowning.file_read_write.server.model.OAuthToken;

import org.jetbrains.annotations.NotNull;

import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelOutboundInvoker;
import io.netty.handler.codec.http.*;
import io.netty.util.CharsetUtil;

public class PostFileCall extends FileCall {

    private final String newFileContent;

    public PostFileCall(OAuthToken oAuthToken,
                        @NotNull String newFileContent,
                        FileReadWriteClient client,
                        HandlerCallback<String> callback) {
        super(oAuthToken, client, callback);
        this.newFileContent = newFileContent;
    }

    @Override
    public void execute() {
        System.out.println("Requesting to post file content...");
        super.execute();
    }

    @Override
    protected FullHttpMessage getRequest(ChannelOutboundInvoker ctx) {
        System.out.println("Posting updated file content...");
        FullHttpMessage message = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1,
                                                             HttpMethod.POST,
                                                             "",
                                                             Unpooled.copiedBuffer(newFileContent,
                                                                                   CharsetUtil.UTF_8));
        message.headers().add(HttpHeaderNames.CONTENT_LENGTH, newFileContent.length());
        return message;
    }
}

File: src/main/scala/com/mattchowning_scala/file_read_write/client/calls/PostFileCall.scala
package com.mattchowning_scala.file_read_write.client.calls

import com.mattchowning_scala.file_read_write.server.model.OAuthToken
import com.mattchowning_scala.file_read_write.client.FileReadWriteClient
import io.netty.buffer.Unpooled
import io.netty.channel.ChannelOutboundInvoker
import io.netty.handler.codec.http._
import io.netty.util.CharsetUtil

class PostFileCall(_oAuthToken: OAuthToken,
                   _client: FileReadWriteClient,
                   successCallback: String => Unit,
                   failureCallback: () => Unit,
                   newFileContent: String)
  extends FileCall(_oAuthToken, _client, successCallback, failureCallback) {

  override def execute() {
    System.out.println("Requesting to post file content...")
    super.execute()
  }

  override protected def makeAuthenticatedRequest(ctx: ChannelOutboundInvoker) {
    val message: FullHttpMessage = new DefaultFullHttpRequest(
      HttpVersion.HTTP_1_1,
      HttpMethod.POST,
      "",
      Unpooled.copiedBuffer(newFileContent, CharsetUtil.UTF_8))
    message.headers.add(
      HttpHeaderNames.CONTENT_LENGTH,
      newFileContent.length).add(HttpHeaderNames.AUTHORIZATION,
      oAuthToken.getEncodedAuthorizationHeader)
    System.out.println("Posting updated file content...")
    ctx.writeAndFlush(message)
  }
}